### PR TITLE
Incremental PSyclone

### PR DESCRIPTION
**Pull request by bblay**
Makes the PSyclone step "incremental", i.e uses prebuilt files and hashes to avoid reprocessing on subsequent runs.
- Analyses x90 with prebuilds, recording kernel dependencies.
- Updates Fortran analyser to record kernel hashes.
- Calls psyclone with prebuilds.

todo:
 - [x] merge smaller chunks we've pulled out from this branch
    - [x] #202 
    - [x] #204 
    - [x] #205
 - [x] go through the todos in this pr
 - [x] don't record and hash the pulled-out names, simply hash the pre-sanitised x90
 - [x] ~discuss: "test cleanup of prebuild files" comment - spin out ticket?~
 - [x] System test `TestPsyclone` passes on its own, fails when run with all system tests
 - [x] tidy up - remove more unnecessary stuff
    - [x] run config changes - into separate pr as they're useful?
    - [x] there's some unnecessary processing of kernel hashes into `used_kernel_hashes` to look into
 - [x] We probably shouldn't be adding hash functions to mutable objects, which I think was just so we can put them into sets. Might want to spin this out into a separate ticket as it extends beyond these psyclone changes.

Merging the analysis result classes pr separately has removed roughly over 1k additions and 600 deletions from this pr.

Also worth pointing out that the sample run configs are not part of the fab source, more just development aids and example ideas for users.

### CODE REVIEW

File: system-tests-new/psyclone/test_psyclone.py [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+
+from fab.build_config import BuildConfig
+from fab.parse.x90 import X90Analyser, AnalysedX90
+from fab.steps.find_source_files import FindSourceFiles
+from fab.steps.grab.folder import GrabFolder
+from fab.steps.preprocess import fortran_preprocessor
+from fab.steps.psyclone import make_parsable_x90, Psyclone, psyclone_preprocessor, tool_available, MpPayload
+from fab.util import file_checksum
+
+SAMPLE_KERNEL = Path(__file__).parent / 'kernel.f90'
```
Comment (MatthewHambley) [inline system-tests-new/psyclone/test_psyclone.py:23]: For extra test coverage in the future we might want to try an algorithm which calls more than one (none built-in) kernel.

File: system-tests-new/psyclone/test_psyclone.py [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+    x90_analyser = X90Analyser()
+    x90_analyser._config = BuildConfig('proj', fab_workspace=tmp_path)
+    x90_analyser._config._prep_output_folders()
+    x90_analyser.run(parsable_x90_path)
+
+    # ensure the files are as expected
+    assert filecmp.cmp(parsable_x90_path, EXPECT_PARSABLE_X90)
+
+    # don't leave this in my git repo
+    unlink(parsable_x90_path)
```
Comment (MatthewHambley) [inline system-tests-new/psyclone/test_psyclone.py:53]: Is this left over debug code or is it useful? If the latter a more indicative comment may help.

**Review by MatthewHambley — CHANGES_REQUESTED** [review]
New functionality looks good. There are also a lot of minor clean-ups in this change which also look okay.

File: run_configs/lfric/lfric_common.py [hunk 3]
```diff
@@ -4,15 +4,13 @@
...
-from fab.steps import Step, check_for_errors
-from fab.steps.preprocess import PreProcessor
-from fab.util import log_or_dot, input_to_output_fpath
+from fab.steps import Step
 from fab.tools import run_command
 
 logger = logging.getLogger('fab')
 
 
+# todo: is this part of psyclone? if so, put  it in the psyclone step module?
```
[the remark this answers is not in the document]
Comment (bblay) [reply run_configs/lfric/lfric_common.py:13]: Ah, ok then I'll leave it where it it and remove this comment, thanks.

File: run_configs/um/build_um.py [hunk 4]
```diff
@@ -128,7 +130,19 @@ def um_atmos_safe_config(revision, two_stage=False):
             ],
         ),
 
-        Analyse(root_symbol='um_main'),
+        Analyse(
+            root_symbol='um_main',
+
+            # depending on environment, fparser2 can fail to parse this file but it does compile.
```
[the remark this answers is not in the document]
Comment (bblay) [reply run_configs/um/build_um.py:136]: I'm sorry but I don't know what environmental changes were causing it to pass/fail, other than assuming it's v.13 Vs v.16,

File: setup.py [hunk 5]
```diff
@@ -20,9 +20,10 @@
     else:
         raise RuntimeError('Cannot determine package version.')
 
+
 tests = ['pytest', 'pytest-cov', 'pytest-mock', 'flake8', 'mypy']
 docs = ['sphinx', 'sphinx_rtd_theme', 'sphinx-autodoc-typehints']
-features = ['GitPython', 'matplotlib']
+features = ['GitPython', 'matplotlib', 'jinja2', 'psyclone==2.1.0']
```
[the remark this answers is not in the document]
Comment (bblay) [reply setup.py:26]: Iirc, we couldn't use >= in the past, a long time ago, because the more recent version was broken/incompatible in some way that I can't remember. I've searched the team chat but can't find mention of it. It's certainly worth going through all the deps and trying the latest versions of things at some point though.

File: source/fab/steps/psyclone.py [hunk 6]
```diff
@@ -0,0 +1,428 @@
...
+from fab.parse.x90 import X90Analyser, AnalysedX90
+from fab.steps import Step, check_for_errors
+from fab.steps.preprocess import PreProcessor
+from fab.util import log_or_dot, input_to_output_fpath, file_checksum, file_walk, TimerLogger, \
+    string_checksum, suffix_filter, by_type, log_or_dot_finish
+
+logger = logging.getLogger(__name__)
+
+
+def tool_available() -> bool:
```
[the remark this answers is not in the document]
Comment (bblay) [reply source/fab/steps/psyclone.py:33]: I'll push back on this unless you really want it, in which case I'm happy to make this change. Reason: the module provides sufficient namespacing, `fab.steps.psyclone.tool_available`, and one shouldn't be forced to construct a build step just to see if the tool is available. For example you might want to fail early at the start of your script.

File: source/fab/steps/psyclone.py [hunk 7]
```diff
@@ -0,0 +1,428 @@
...
+    try:
+        run_command(['psyclone', '-h'])
+    except (RuntimeError, FileNotFoundError):
+        return False
+    return True
+
+
+# todo: should this be part of the psyclone step?
+def psyclone_preprocessor(set_um_physics=False):
```
[the remark this answers is not in the document]
Comment (bblay) [reply source/fab/steps/psyclone.py:42]: Quite right, good point.

File: source/fab/steps/psyclone.py [hunk 8]
```diff
@@ -0,0 +1,428 @@
...
+        Some kernels, such as `setval_c`, are
+        `PSyclone built-ins <https://github.com/stfc/PSyclone/blob/ebb7f1aa32a9377da6ccc1ec04eec4adbc1e0a0a/src/
+        psyclone/domain/lfric/lfric_builtins.py#L2136>`_.
+        They will not appear in use statements and can be ignored.
+
+        The Psyclone and Analyse steps both use the generic Fortran analyser, which recognises Psyclone kernel metadata.
+        The Analysis step must come after this step because it needs to analyse the fortran we create.
+
+        """
+        mp_payload = MpPayload()
```
[the remark this answers is not in the document]
Comment (bblay) [reply source/fab/steps/psyclone.py:173]: I think yes, there's a general pattern here for passing data to child processes without using protected member attributes, if that's what you mean?

File: system-tests-new/psyclone/test_psyclone.py (hunk 1 above)
Comment (bblay) [reply system-tests-new/psyclone/test_psyclone.py:23]: Agreed. Added a todo for that so it's not forgotten.

File: source/fab/steps/psyclone.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (MatthewHambley) [reply source/fab/steps/psyclone.py:33]: The potential issue I see is where there is more than one tool. Consider:

```python
from fab.steps.psyclone import tool_available
from fab.steps.fortran import tool_available
```
To make this work would require renaming syntax which is extra work for the developer.

There is no need to instantiate a class to access its static methods so:

```python
from fab.steps.psyclone import Psyclone
...
if Psyclone.tool_available():
```

Things work as they stand so I don't intend to block this change on this matter but there is a discussion to be had for a potential future change.

**Review by MatthewHambley — CHANGES_REQUESTED** [review]
This change highlights a need to really interrogate all our dependencies and ensure we are specifying them in the most appropriate way. e.g. should we be requiring only a particular version of PSyclone when I can guarantee that the LFRic project will be upgrading at some point.

This is a matter for a different ticket.

The only outstanding concern is really just a request for comment at this point.

File: source/fab/steps/psyclone.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (bblay) [reply source/fab/steps/psyclone.py:33]: Your example is convincing, I might as well make that change now.

**Comment by bblay** [comment]
I've [raised a ticket](https://github.com/orgs/metomi/projects/1/views/1?filterQuery=version&pane=issue&itemId=21331971) to think about dependency versions.
